### PR TITLE
Calculating fee from coins

### PR DESCRIPTION
**Pull request by micahriggan**

### CODE REVIEW

**Comment by skddc** [comment]
Interesting. How is the fee calculated exactly?

**Comment by micahriggan** [comment]
Sum of spent coins minus the sum of created coins for a transaction

**Comment by skddc** [comment]
Oh, I misread. I thought this was about fee estimation, hence my confusion when looking at the code. Thanks.

File: packages/bitcore-node/src/models/transaction.ts [hunk 1]
```diff
@@ -104,41 +104,49 @@ export class Transaction extends BaseModel<ITransaction> {
...
+      }
+      return agg;
+    }, {});
 
     let txOps = txs.map((tx, index) => {
-      let wallets = new Array<ObjectID>();
-      if (initialSyncComplete) {
-        for (let wallet of mintWallets.concat(spentWallets).filter(wallet => wallet._id === txids[index])) {
-          for (let walletMatch of wallet.wallets) {
-            if (!wallets.find(wallet => wallet.toHexString() === walletMatch.toHexString())) {
```
Comment (matiu) [inline packages/bitcore-node/src/models/transaction.ts]: This 'unique' check was implemented on the old code, but I can't see it on the new version, Am I missing something?

Comment (micahriggan) [reply packages/bitcore-node/src/models/transaction.ts]: No, that should probably be there. I thought it was a Set on the mongo sid

File: packages/bitcore-node/src/models/transaction.ts [hunk 2]
```diff
@@ -104,41 +105,50 @@ export class Transaction extends BaseModel<ITransaction> {
...
-
-    if (initialSyncComplete) {
-      mintWallets = await CoinModel.collection
-        .aggregate<TaggedCoin>([
-          { $match: { mintTxid: { $in: txids }, chain, network } },
-          { $unwind: '$wallets' },
-          { $group: { _id: '$mintTxid', wallets: { $addToSet: '$wallets' } } }
-        ])
-        .toArray();
+    const mintedPromise = CoinModel.collection.find({ mintTxid: { $in: txids }, chain, network }).toArray();
```
Comment (nitsujlangston) [inline packages/bitcore-node/src/models/transaction.ts]: use mintops from previous step

File: packages/bitcore-node/src/models/transaction.ts [hunk 3]
```diff
@@ -104,41 +105,50 @@ export class Transaction extends BaseModel<ITransaction> {
...
+    type CoinGroup = { [txid: string]: { total: number; wallets: Array<ObjectID> } };
+    const groupedMints = minted.reduce<CoinGroup>((agg, current) => {
+      if (!agg[current.mintTxid]) {
+        agg[current.mintTxid] = {
+          total: current.value,
+          wallets: current.wallets || []
+        };
+      } else {
+        agg[current.mintTxid].total += current.value;
+        agg[current.mintTxid].wallets.push(...current.wallets);
```
Comment (nitsujlangston) [inline packages/bitcore-node/src/models/transaction.ts]: wallets would come from mintOps

File: packages/bitcore-node/src/models/transaction.ts [hunk 4]
```diff
@@ -104,41 +105,50 @@ export class Transaction extends BaseModel<ITransaction> {
...
-    if (initialSyncComplete) {
-      mintWallets = await CoinModel.collection
-        .aggregate<TaggedCoin>([
-          { $match: { mintTxid: { $in: txids }, chain, network } },
-          { $unwind: '$wallets' },
-          { $group: { _id: '$mintTxid', wallets: { $addToSet: '$wallets' } } }
-        ])
-        .toArray();
+    const mintedPromise = CoinModel.collection.find({ mintTxid: { $in: txids }, chain, network }).toArray();
+    const spentPromise = CoinModel.collection.find({ spentTxid: { $in: txids }, chain, network }).toArray();
```
Comment (nitsujlangston) [inline packages/bitcore-node/src/models/transaction.ts]: { batchSize: 100 }

**Pull request merged**